File: sdks/go/domain/validation_test.go
package domain

import (
	"fmt"
	"strings"
	"testing"
)

func TestValidateControlPointKey(t *testing.T) {
	if _, err := ValidateControlPointKey(""); err == nil || err.Kind != KindFlagNotFound {
		t.Errorf("empty key: got %v, want FlagNotFound", err)
	}
	if _, err := ValidateControlPointKey(strings.Repeat("k", 257)); err == nil || err.Kind != KindFlagNotFound {
		t.Errorf("too long: got %v, want FlagNotFound", err)
	}
	if _, err := ValidateControlPointKey("bad\x00key"); err == nil || err.Kind != KindFlagNotFound {
		t.Errorf("control char: got %v, want FlagNotFound", err)
	}
	if v, err := ValidateControlPointKey("good-key"); err != nil || v != "good-key" {
		t.Errorf("valid key: got (%q, %v)", v, err)
	}
}

func TestValidateDefaultValue(t *testing.T) {
	cases := []struct {
		expected FlagType
		value    any
		wantOK   bool
	}{
		{FlagTypeBoolean, true, true},
		{FlagTypeBoolean, "nope", false},
		{FlagTypeString, "s", true},
		{FlagTypeString, 1, false},
		{FlagTypeNumber, 1.5, true},
		{FlagTypeNumber, 5, true},
		{FlagTypeNumber, "5", false},
		{FlagTypeObject, map[string]any{"a": 1}, true},
		{FlagTypeObject, []any{1, 2}, true},
		{FlagTypeObject, "not-an-object", false},
		{FlagTypeBoolean, nil, false},
	}
	for _, tc := range cases {
		_, err := ValidateDefaultValue(tc.expected, tc.value)
		gotOK := err == nil
		if gotOK != tc.wantOK {
			t.Errorf("ValidateDefaultValue(%s, %#v) ok=%v, want %v (err=%v)", tc.expected, tc.value, gotOK, tc.wantOK, err)
		}
		if !tc.wantOK && err.Kind != KindTypeMismatch {
			t.Errorf("expected TypeMismatch, got %s", err.Kind)
		}
	}
}

func TestValidateBounds(t *testing.T) {
	limits := DefaultLimits()

	t.Run("attribute count", func(t *testing.T) {
		attrs := map[string]any{}
		for i := 0; i < 129; i++ {
			attrs[fmt.Sprintf("a%03d", i)] = i
		}
		_, err := ValidateContext(EvaluationContext{TargetingKey: "k", Attributes: attrs}, limits, false)
		if err == nil || err.Message != "context exceeds maximum attribute count" {
			t.Errorf("got %v, want attribute-count violation", err)
		}
	})

	t.Run("key size", func(t *testing.T) {
		attrs := map[string]any{strings.Repeat("K", 257): "x"}
		_, err := ValidateContext(EvaluationContext{TargetingKey: "k", Attributes: attrs}, limits, false)
		if err == nil || err.Message != "context key exceeds maximum size" {
			t.Errorf("got %v, want key-size violation", err)
		}
	})

	t.Run("value size", func(t *testing.T) {
		attrs := map[string]any{"blob": strings.Repeat("B", 4097)}
		_, err := ValidateContext(EvaluationContext{TargetingKey: "k", Attributes: attrs}, limits, false)
		if err == nil || err.Message != "context value exceeds maximum size" {
			t.Errorf("got %v, want value-size violation", err)
		}
	})

	t.Run("nesting depth", func(t *testing.T) {
		deep := any(true)
		for i := 0; i < 8; i++ {
			deep = map[string]any{"d": deep}
		}
		_, err := ValidateContext(EvaluationContext{TargetingKey: "k", Attributes: map[string]any{"deep": deep}}, limits, false)
		if err == nil || err.Message != "context exceeds maximum nesting depth" {
			t.Errorf("got %v, want depth violation", err)
		}
	})

	t.Run("serialized size", func(t *testing.T) {
		attrs := map[string]any{}
		for i := 0; i < 20; i++ {
			attrs["p"+string(rune('a'+i))] = strings.Repeat("X", 4000)
		}
		_, err := ValidateContext(EvaluationContext{TargetingKey: "k", Attributes: attrs}, limits, false)
		if err == nil || err.Message != "serialized context exceeds maximum size" {
			t.Errorf("got %v, want serialized-size violation", err)
		}
	})

	t.Run("depth six allowed", func(t *testing.T) {
		v := any(true)
		for i := 0; i < 5; i++ {
			v = map[string]any{"d": v}
		}
		if _, err := ValidateContext(EvaluationContext{TargetingKey: "k", Attributes: map[string]any{"v": v}}, limits, false); err != nil {
			t.Errorf("depth 6 should be allowed, got %v", err)
		}
	})
}

func TestValidateReservedKeys(t *testing.T) {
	cases := []map[string]any{
		{"targetingKey": "dup"},
		{"kind": "user"},
		{"fireweave.anything": 1},
		{"fireweave.evaluationContexts": []any{"a"}},
		{AttrGroups: "not-a-map"},
		{AttrGroupProperties: []any{"x"}},
	}
	for _, attrs := range cases {
		_, err := ValidateContext(EvaluationContext{TargetingKey: "k", Attributes: attrs}, DefaultLimits(), false)
		if err == nil || err.Kind != KindInvalidContext {
			t.Errorf("attrs %v: got %v, want InvalidContext", attrs, err)
		}
	}
}

func TestValidateFireweaveGroupsCarveOut(t *testing.T) {
	attrs := map[string]any{
		AttrGroups:          map[string]any{"organization": "org_1"},
		AttrGroupProperties: map[string]any{"organization": map[string]any{"plan": "enterprise"}},
		"email_domain":      "example.com",
	}
	if _, err := ValidateContext(EvaluationContext{TargetingKey: "k", Attributes: attrs}, DefaultLimits(), false); err != nil {
		t.Fatalf("canonical carve-out keys must validate, got %v", err)
	}
}

func TestValidateTargetingKeyRequired(t *testing.T) {
	_, err := ValidateContext(EvaluationContext{Attributes: map[string]any{"plan": "pro"}}, DefaultLimits(), true)
	if err == nil || !err.TargetingKeyMissing {
		t.Fatalf("got %v, want TargetingKeyMissing error", err)
	}
	if err.Message != "targeting key missing" {
		t.Errorf("message = %q", err.Message)
	}
	if _, e := ValidateContext(EvaluationContext{Attributes: map[string]any{"plan": "pro"}}, DefaultLimits(), false); e != nil {
		t.Errorf("not required: got %v, want nil", e)
	}
}

func TestValidateInitOptions(t *testing.T) {
	cases := []struct {
		name           string
		mode           Mode
		apiKey, apiURL string
		wantOK         bool
	}{
		{"mode absent", "", "", "", false},
		{"mode unrecognised", Mode("bogus"), "", "", false},
		{"remote missing apiKey", ModeRemote, "", "https://app-server.fireweave.ai", false},
		{"remote missing apiUrl", ModeRemote, "project-api-key_test", "", false},
		{"remote blank apiKey", ModeRemote, "   ", "https://app-server.fireweave.ai", false},
		{"remote ok", ModeRemote, "project-api-key_test", "https://app-server.fireweave.ai", true},
		{"local with apiKey", ModeLocal, "project-api-key_test", "", false},
		{"local with apiUrl", ModeLocal, "", "https://app-server.fireweave.ai", false},
		{"local with both", ModeLocal, "project-api-key_test", "https://app-server.fireweave.ai", false},
		{"local ok", ModeLocal, "", "", true},
		{"local blank not supplied", ModeLocal, "  ", "  ", true},
	}
	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			err := ValidateInitOptions(tc.mode, tc.apiKey, tc.apiURL)
			gotOK := err == nil
			if gotOK != tc.wantOK {
				t.Errorf("got ok=%v (err=%v), want %v", gotOK, err, tc.wantOK)
			}
			if !tc.wantOK && err.Kind != KindConfiguration {
				t.Errorf("expected Configuration, got %s", err.Kind)
			}
		})
	}
}

File: sdks/dart/test/validation_test.dart
import 'package:fireweave/fireweave.dart';
import 'package:test/test.dart';

extension on Validated {
  ErrorKind? get errorKind => error?.kind;
}

void main() {
  group('validateControlPointKey', () {
    test('must be non-empty', () {
      expect(validateControlPointKey('').isValid, isFalse);
      expect(validateControlPointKey('ok').isValid, isTrue);
    });

    test('length is counted in characters', () {
      expect(validateControlPointKey('k' * 257).isValid, isFalse);
      expect(validateControlPointKey('k' * 256).isValid, isTrue);
    });

    test('rejects control characters', () {
      expect(validateControlPointKey('badkey').isValid, isFalse);
      expect(
        validateControlPointKey('badkey').errorKind,
        ErrorKind.flagNotFound,
      );
    });
  });

  group('validateDefaultValue', () {
    test('type mismatch', () {
      expect(
        validateDefaultValue(FlagType.boolean, 'not-a-bool').errorKind,
        ErrorKind.typeMismatch,
      );
      expect(validateDefaultValue(FlagType.boolean, true).isValid, isTrue);
      expect(validateDefaultValue(FlagType.number, 2).isValid, isTrue);
      expect(validateDefaultValue(FlagType.number, 2.5).isValid, isTrue);
      expect(
        validateDefaultValue(FlagType.object, <Object?>[]).isValid,
        isTrue,
      );
      expect(validateDefaultValue(FlagType.object, 'x').isValid, isFalse);
    });
  });

  group('validateTargetingKey', () {
    test('required and missing', () {
      final result = validateTargetingKey(null, required: true);
      expect(result.error?.openFeatureErrorCode, 'TARGETING_KEY_MISSING');
      expect(validateTargetingKey(null, required: false).isValid, isTrue);
      expect(validateTargetingKey('x', required: true).isValid, isTrue);
      expect(validateTargetingKey('', required: true).isValid, isFalse);
    });
  });

  group('validateContext', () {
    Validated check(
      EvaluationContext ctx, {
      Set<String> reserved = defaultReservedAttributeKeys,
      bool requireTargetingKey = false,
      ContextLimits limits = defaultContextLimits,
    }) => validateContext(
      ctx,
      limits: limits,
      reservedKeys: reserved,
      requireTargetingKey: requireTargetingKey,
    );

    test('reserved keys rejected', () {
      final ctx = EvaluationContext(
        targetingKey: 't',
        attributes: {'targetingKey': 'dup'},
      );
      expect(check(ctx).errorKind, ErrorKind.invalidContext);
      expect(check(ctx).error?.message, 'invalid evaluation context');
    });

    test('fireweave carve-out keys allowed, others rejected', () {
      final ok = EvaluationContext(
        targetingKey: 't',
        attributes: {
          'fireweave.groups': {'organization': 'org_1'},
        },
      );
      expect(check(ok, reserved: {}).isValid, isTrue);
      final bad = EvaluationContext(
        targetingKey: 't',
        attributes: {
          'fireweave.evaluationContexts': ['production'],
        },
      );
      expect(check(bad, reserved: {}).isValid, isFalse);
    });

    test('nesting depth exceeded', () {
      Object? nested = {'d9': true};
      for (final name in ['d8', 'd7', 'd6', 'd5', 'd4', 'd3', 'd2', 'd1']) {
        nested = {name: nested};
      }
      final ctx = EvaluationContext(
        targetingKey: 't',
        attributes: (nested as Map).cast<String, Object?>(),
      );
      expect(
        check(ctx).error?.message,
        'context exceeds maximum nesting depth',
      );
    });

    test('a flat context at depth 1 and a nested map at depth 2 are fine', () {
      final ctx = EvaluationContext(
        targetingKey: 't',
        attributes: {
          'plan': 'pro',
          'nested': {'a': 1},
        },
      );
      expect(check(ctx).isValid, isTrue);
    });

    test('attribute count exceeded', () {
      final attrs = {for (var i = 0; i < 200; i += 1) 'a$i': i};
      final ctx = EvaluationContext(targetingKey: 't', attributes: attrs);
      expect(
        check(ctx).error?.message,
        'context exceeds maximum attribute count',
      );
    });

    test('key size exceeded (utf-8 bytes, at any depth)', () {
      final ctx = EvaluationContext(
        targetingKey: 't',
        attributes: {
          'outer': {'k' * 257: 'x'},
        },
      );
      expect(check(ctx).error?.message, 'context key exceeds maximum size');
    });

    test('value size exceeded (utf-8 bytes)', () {
      final ctx = EvaluationContext(
        targetingKey: 't',
        attributes: {'blob': 'é' * 2100},
      );
      expect(check(ctx).error?.message, 'context value exceeds maximum size');
    });

    test('serialized size exceeded', () {
      final attrs = {for (var i = 0; i < 40; i += 1) 'p$i': 'X' * 2000};
      final ctx = EvaluationContext(targetingKey: 't', attributes: attrs);
      expect(
        check(ctx).error?.message,
        'serialized context exceeds maximum size',
      );
    });

    test('targeting key required', () {
      final ctx = EvaluationContext(attributes: {'plan': 'pro'});
      final result = check(ctx, requireTargetingKey: true);
      expect(result.error?.openFeatureErrorCode, 'TARGETING_KEY_MISSING');
      expect(result.errorKind, ErrorKind.invalidContext);
    });
  });

  group('validateInitOptions', () {
    test('mode absent is Configuration / PROVIDER_FATAL', () {
      final result = validateInitOptions(mode: null);
      expect(result.errorKind, ErrorKind.configuration);
      expect(result.error?.openFeatureErrorCode, 'PROVIDER_FATAL');
    });

    test('remote requires credentials', () {
      expect(
        validateInitOptions(mode: Mode.remote, apiUrl: 'https://x').isValid,
        isFalse,
      );
      expect(
        validateInitOptions(
          mode: Mode.remote,
          apiKey: 'key',
          apiUrl: '  ',
        ).isValid,
        isFalse,
      );
      expect(
        validateInitOptions(
          mode: Mode.remote,
          apiKey: 'key',
          apiUrl: 'https://x',
        ).isValid,
        isTrue,
      );
    });

    test('local rejects stray credentials', () {
      expect(
        validateInitOptions(mode: Mode.local, apiKey: 'key').isValid,
        isFalse,
      );
      expect(validateInitOptions(mode: Mode.local).isValid, isTrue);
    });
  });
}
